File: Cs/Acme/Infrastructure/InMemory/Domain/InMemoryUsers.cs
using Cs.Acme.Domain;
using Cs.Acme.Domain.Model;
using System.Collections.Generic;

namespace Cs.Acme.Infrastructure.InMemory.Domain
{
    public class InMemoryUsers : IUsers
    {
        private readonly Dictionary<string, User> _users;

        public InMemoryUsers()
        {
            _users = new Dictionary<string, User>();
        }

        public void Add(User user)
        {
            _users[user.Email] = user;
        }

        public bool Exists(string email)
        {
            return _users.ContainsKey(email);
        }

        public User GetByEmail(string email)
        {
            return _users[email];
        }

        public void Remove(string email)
        {
            _users.Remove(email);
        }
    }
}

File: dart/lib/acme/infrastructure/inmemory/domain/inmemoryusers.dart
import 'dart:collection';

import 'package:dart/acme/domain/model/user.dart';
import 'package:dart/acme/domain/users.dart';

class InMemoryUsers implements Users {
  final Map<String, User> _users;

  InMemoryUsers() : _users = HashMap<String, User>();

  @override
  void add(User user) {
    _users[user.email] = user;
  }

  @override
  bool exists(String email) {
    return _users.containsKey(email);
  }

  @override
  User getByEmail(String email) {
    return _users[email];
  }

  @override
  void remove(String email) {
    _users.remove(email);
  }
}